src/shims/thread.rs

```diff
@@ -0,0 +1,122 @@
+        this.write_scalar(
+            Scalar::from_uint(new_thread_id.to_u32(), thread_info_place.layout.size),
+            thread_info_place.into(),
+        )?;
+
+        let fn_ptr = this.read_scalar(start_routine)?.not_undef()?;
+        let instance = this.memory.get_fn(fn_ptr)?.as_instance()?;
+
+        let func_arg = this.read_immediate(arg)?;
+
+        // Note: the returned value is currently ignored (see the FIXME in
+        // pthread_join below) because the Rust standard library does not use
+        // it.
+        let ret_place =
+            this.allocate(this.layout_of(this.tcx.types.usize)?, MiriMemoryKind::Machine.into());
+
+        this.call_function(
+            instance,
+            &[*func_arg],
+            Some(ret_place.into()),
+            StackPopCleanup::None { cleanup: true },
+        )?;
+
+        this.set_active_thread(old_thread_id)?;
+
+        Ok(0)
+    }
+
+    fn pthread_join(
+        &mut self,
+        thread: OpTy<'tcx, Tag>,
+        retval: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.is_null(this.read_scalar(retval)?.not_undef()?)? {
+            // FIXME: implement reading the thread function's return place.
+            throw_unsup_format!("Miri supports pthread_join only with retval==NULL");
+        }
+
+        let thread_id = this.read_scalar(thread)?.to_machine_usize(this)?;
+        this.join_thread(thread_id.try_into().expect("thread ID should fit in u32"))?;
+
+        Ok(0)
+    }
+
+    fn pthread_detach(&mut self, thread: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let thread_id = this.read_scalar(thread)?.to_machine_usize(this)?;
+        this.detach_thread(thread_id.try_into().expect("thread ID should fit in u32"))?;
+
+        Ok(0)
+    }
+
+    fn pthread_self(&mut self, dest: PlaceTy<'tcx, Tag>) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        let thread_id = this.get_active_thread()?;
+        this.write_scalar(Scalar::from_uint(thread_id.to_u32(), dest.layout.size), dest)
+    }
+
+    fn prctl(
+        &mut self,
+        option: OpTy<'tcx, Tag>,
+        arg2: OpTy<'tcx, Tag>,
+        _arg3: OpTy<'tcx, Tag>,
+        _arg4: OpTy<'tcx, Tag>,
+        _arg5: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let option = this.read_scalar(option)?.to_i32()?;
+        if option == this.eval_libc_i32("PR_SET_NAME")? {
+            let address = this.read_scalar(arg2)?.not_undef()?;
+            let name = this.memory.read_c_str(address)?.to_owned();
+            this.set_active_thread_name(name)?;
+        } else if option == this.eval_libc_i32("PR_GET_NAME")? {
+            let address = this.read_scalar(arg2)?.not_undef()?;
+            let name = this.get_active_thread_name()?.to_vec();
+            this.memory.write_bytes(address, name)?;
```

Comment:
`read_c_str` above returns the bytes *without* the NULL terminator, but here you need to write a NULL terminator. (`name.into_iter().chain(iter::once(0u8))` is what we usually do for that).

Also, given the docs say that the name is max. 16 bytes including the terminator, maybe add an `assert!` that ensures this.